File: apps/control-api/app/routers/discovery_origin.py
"""Read-only origin projection; never infer a framework/skill relationship."""

from fastapi import APIRouter, Depends, HTTPException, Response
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.db import get_session
from app.discovery_identity import asset_evidence_device_filter
from app.framework_source_view import project_framework_sources
from app.models import AgentAsset, EdgeAgent, Environment, Evidence, RoleSkillSelectionObservation
from app.security import Identity, ensure_permission, get_identity

router = APIRouter(tags=["inventory"])


@router.get("/api/v1/agents/{asset_id}/framework-source")
def framework_source(
    asset_id: str,
    response: Response,
    session: Session = Depends(get_session),
    identity: Identity = Depends(get_identity),
):
    asset = session.scalar(select(AgentAsset).where(
        AgentAsset.id == asset_id, AgentAsset.tenant_id == identity.tenant_id,
    ))
    if asset is None:
        raise HTTPException(404, "not_found")
    ensure_permission(identity, "agent:read")
    ensure_permission(identity, "env:read")
    response.headers["Cache-Control"] = "no-store"
    return project_framework_sources(session, identity.tenant_id, [asset])[asset.id]


@router.get("/api/v1/agents/{asset_id}/skill-selections")
def skill_selections(
    asset_id: str,
    response: Response,
    session: Session = Depends(get_session),
    identity: Identity = Depends(get_identity),
):
    asset = session.scalar(select(AgentAsset).where(
        AgentAsset.id == asset_id, AgentAsset.tenant_id == identity.tenant_id,
    ))
    if asset is None:
        raise HTTPException(status_code=404, detail="not_found")
    ensure_permission(identity, "agent:read")
    ensure_permission(identity, "env:read")
    rows = session.execute(
        select(RoleSkillSelectionObservation, EdgeAgent)
        .join(EdgeAgent, RoleSkillSelectionObservation.edge_agent_id == EdgeAgent.id)
        .join(Environment, EdgeAgent.environment_id == Environment.id)
        .where(
            RoleSkillSelectionObservation.tenant_id == identity.tenant_id,
            RoleSkillSelectionObservation.asset_id == asset.id,
            RoleSkillSelectionObservation.edge_agent_id == asset.discovery_scope,
            Environment.tenant_id == identity.tenant_id,
        )
        .order_by(RoleSkillSelectionObservation.received_at.desc(), RoleSkillSelectionObservation.id.desc())
        .limit(101)
    ).all()
    response.headers["Cache-Control"] = "no-store"
    return {
        "schema_version": "enterprise-role-skill-observations/v1",
        "asset_id": asset.id,
        "status": "historical_declarations" if rows else "no_recorded_declaration",
        "relationship_status": "unresolved",
        "effective_permissions": None,
        "observations_truncated": len(rows) > 100,
        "observations": [{
            "id": row.id,
            "device": {"id": edge.id, "revoked": edge.revoked_at is not None},
            "task_id": row.task_id,
            "batch_digest": row.batch_digest,
            "selection": row.selection,
            "source_evidence": row.source_evidence,
            "observed_at": row.observed_at.isoformat() + "Z",
            "received_at": row.received_at.isoformat() + "Z",
        } for row, edge in rows[:100]],
    }


@router.get("/api/v1/agents/{asset_id}/discovery-origin")
def discovery_origin(
    asset_id: str,
    response: Response,
    session: Session = Depends(get_session),
    identity: Identity = Depends(get_identity),
):
    asset = session.scalar(
        select(AgentAsset).where(AgentAsset.id == asset_id, AgentAsset.tenant_id == identity.tenant_id)
    )
    if asset is None:
        raise HTTPException(status_code=404, detail="not_found")
    ensure_permission(identity, "agent:read")
    ensure_permission(identity, "env:read")
    response.headers["Cache-Control"] = "no-store"
    result = {
        "schema_version": "enterprise-discovery-origin/v1",
        "asset_id": asset.id,
        "status": "legacy_unresolved" if asset.discovery_scope == "legacy" else "source_unavailable",
        "environment": None,
        "device": None,
        "reported_framework": asset.framework,
        "assigned_role": asset.role,
        "observations": [],
        "observations_truncated": False,
    }
    if asset.discovery_scope == "legacy":
        return result
    source = session.execute(
        select(EdgeAgent, Environment)
        .join(Environment, EdgeAgent.environment_id == Environment.id)
        .where(EdgeAgent.id == asset.discovery_scope, Environment.tenant_id == identity.tenant_id)
    ).first()
    if source is None:
        return result
    edge, environment = source
    result["status"] = "device_bound"
    result["environment"] = {"id": environment.id, "name": environment.name}
    result["device"] = {"id": edge.id, "identity": edge.device_identity, "revoked": edge.revoked_at is not None}
    observations = list(
        session.scalars(
            select(Evidence)
            .where(
                Evidence.tenant_id == identity.tenant_id,
                Evidence.evidence_id.in_(asset.evidence_ids or []) | (Evidence.subject_ref == asset.id),
                asset_evidence_device_filter(asset),
            )
            .order_by(Evidence.observed_at.desc(), Evidence.id)
            .limit(201)
        )
    )
    result["observations_truncated"] = len(observations) > 200
    result["observations"] = [
        {
            "observation_id": row.id,
            "evidence_id": row.evidence_id,
            "content_hash": row.content_hash,
            "observed_at": row.observed_at.isoformat() + "Z",
        }
        for row in observations[:200]
    ]
    return result

File: apps/web/src/api/discoveryOrigin.ts
import { get } from './client';

export interface DiscoveryOrigin {
  schema_version: 'enterprise-discovery-origin/v1';
  asset_id: string;
  status: 'device_bound' | 'legacy_unresolved' | 'source_unavailable';
  environment: { id: string; name: string } | null;
  device: { id: string; identity: string; revoked: boolean } | null;
  reported_framework: string;
  assigned_role: string | null;
  observations: { observation_id: string; evidence_id: string; content_hash: string; observed_at: string }[];
  observations_truncated: boolean;
}
const object = (v: unknown): v is Record<string, unknown> => !!v && typeof v === 'object' && !Array.isArray(v);
const text = (v: unknown): v is string => typeof v === 'string' && v.length > 0 && v.length <= 4096;

export function isDiscoveryOrigin(v: unknown, assetId: string): v is DiscoveryOrigin {
  if (!object(v) || v.schema_version !== 'enterprise-discovery-origin/v1' || v.asset_id !== assetId
    || !text(v.reported_framework) || !(v.assigned_role === null || typeof v.assigned_role === 'string')
    || typeof v.observations_truncated !== 'boolean' || !Array.isArray(v.observations) || v.observations.length > 200) return false;
  if (v.status === 'device_bound') {
    if (!object(v.environment) || !text(v.environment.id) || !text(v.environment.name)
      || !object(v.device) || !text(v.device.id) || !text(v.device.identity) || typeof v.device.revoked !== 'boolean') return false;
  } else if (!['legacy_unresolved', 'source_unavailable'].includes(String(v.status))
    || v.environment !== null || v.device !== null || v.observations.length !== 0 || v.observations_truncated) return false;
  return v.observations.every(o => object(o) && text(o.observation_id) && text(o.evidence_id)
    && typeof o.content_hash === 'string' && /^[a-f0-9]{64}$/.test(o.content_hash)
    && text(o.observed_at) && Number.isFinite(Date.parse(o.observed_at)))
    && new Set(v.observations.map(o => o.observation_id)).size === v.observations.length;
}

export async function getDiscoveryOrigin(assetId: string): Promise<DiscoveryOrigin> {
  const value = await get<unknown>(`/agents/${encodeURIComponent(assetId)}/discovery-origin`);
  if (!isDiscoveryOrigin(value, assetId)) throw new Error('资产来源响应无法核验');
  return value;
}
